course-material/interactive-queries-and-parameterization/introduction-to-ipywidgets.md

```diff
@@ -1,15 +1,376 @@
+
+Numeric widgets provide further flexibility over basic data types. For example, the `IntSlider` and the `FloatSlider`, the simplest numeric widgets, can be employed to filter <b>integer</b> and <b>float</b> values respectively within a range (`min` and `max`) of your choice along with a `step` size and the `value` at which the slider is initialized. Moreover, we can create a bounded slider, indicated by the `<=` operator, or an unbounded slider, indicated by the `>=` operator, to filter the dataset within or outside our range of values respectively.
+
+There are several other arguments, which can be found [here](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html#IntSlider), that can be passed into these sliders.
+
+An example for the `IntSlider` is as follows:
+
+```{code-cell} ipython3
+duration_lower_bound = widgets.IntSlider(min=5, max=3000, step=500, value=1500)
+
+%sql --interact duration_lower_bound SELECT * FROM bank WHERE duration <= {{duration_lower_bound}} LIMIT 5
+```
+
+<b>Note</b>: Other Numeric Widgets can be found [here](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html#floatlogslider).
+
+### Selection Widgets
+
+There are several widgets that can be used to display single selection lists, and two that can be used to select multiple values. All inherit from the same base class. You can specify the <b>enumeration of selectable options by passing a list</b>. Options are either (label, value) pairs, or simply values for which the labels are derived by calling `str`.
+
+#### `RadioButtons`
+
+The `RadioButtons` widget displays a list of options, of which <b>exactly one</b> can be selected. The user can <b>select one of the options</b> by clicking on the radio button. The current selected value can be accessed from the `value` attribute, which is by default the label of the selected option.
+
+An example is as follows:
+
+```{code-cell} ipython3
+outcome_selection = widgets.RadioButtons(
+    options=["failure", "other", "success", "unknown"],
+    description="Outcome",
+    disabled=False,
+)
+```
+
+```{code-cell} ipython3
+%%sql --interact outcome_selection 
+SELECT * FROM bank 
+WHERE poutcome == '{{outcome_selection}}' 
+LIMIT 5
+```
+
+<b>Note</b>: Other Selection Widgets can be found [here](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html#selection-widgets).
+
+### Complete Example
+
+Now, we can demonstrate a way to combine multiple `ipywidgets` to make a more complex interactive SQL Clause, inclusive of different data types and widgets.
+
+The code chunk below initializes multiple `ipywidgets`: <b>radio buttons</b> and <b>multiple selection</b> for the categorical variables `poutcome` and `loan` respectively and a <b>basic unbounded slider</b> for the numeric variable `duration`. The `show_limit` variable is basic data type that creates a slider by default and is used to limit the number of rows, in this case between 0 to 10 with a step size of 1.
```

Comment:
The `show_limit` variable is basic data type -> The `show_limit` varialbe is **a** basic datatype